sdk/clientcore/http-netty4/src/main/java/io/clientcore/http/netty4/implementation/CoreHttpProxyHandler.java

```diff
@@ -0,0 +1,256 @@
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.http.DefaultFullHttpRequest;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpClientCodec;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpUtil;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.handler.proxy.HttpProxyHandler;
+import io.netty.handler.proxy.ProxyConnectException;
+import io.netty.handler.proxy.ProxyHandler;
+
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static io.clientcore.core.utils.AuthUtils.parseAuthenticationOrAuthorizationHeader;
+import static io.clientcore.http.netty4.implementation.NettyUtility.createCodec;
+
+/**
+ * This class handles authorizing requests being sent through a proxy which require authentication.
+ */
+public final class CoreHttpProxyHandler extends ProxyHandler {
+    private static final String VALIDATION_ERROR_TEMPLATE = "The '%s' returned in the 'Proxy-Authentication-Info' "
+        + "header doesn't match the value sent in the 'Proxy-Authorization' header. Sent: %s, received: %s.";
+
+    private static final String PROXY_AUTHENTICATION_INFO = "Proxy-Authentication-Info";
+    private static final HttpHeaderName PROXY_AUTHENTICATION_INFO_NAME
+        = HttpHeaderName.fromString(PROXY_AUTHENTICATION_INFO);
+
+    private static final String NONE = "none";
+    private static final String HTTP = "http";
+
+    private static final String CNONCE = "cnonce";
+    private static final String NC = "nc";
+
+    // CoreHttpProxyHandler will be created for each network request that is using proxy, use a static logger.
+    private static final ClientLogger LOGGER = new ClientLogger(CoreHttpProxyHandler.class);
+
+    private final ProxyOptions proxyOptions;
+    private final AtomicReference<List<AuthenticateChallenge>> proxyChallengeHolder;
+    private final HttpClientCodec codec;
+
+    private String authScheme = null;
+    private String lastUsedAuthorizationHeader;
+
+    private HttpResponseStatus status;
+    private HttpHeaders innerHeaders;
+
+    /**
+     * Creates an instance of HttpProxyHandler.
+     *
+     * @param proxyOptions The configured {@link ProxyOptions} that will dictate the behavior of proxying.
+     * @param proxyChallengeHolder An {@link AtomicReference} containing parsed {@code Proxy-Authenticate} challenges.
+     */
+    public CoreHttpProxyHandler(ProxyOptions proxyOptions,
+        AtomicReference<List<AuthenticateChallenge>> proxyChallengeHolder) {
```

Comment:
nit `proxyChallengeHolder` sure sounds like an instance of a custom type, not an `AtomicReference<List<>>`. `proxyChallenges`? 

Now that I think about it, it also seems somewhat odd that we're being given an `AtomicReference` here? Does something outside this class modify the reference at some point?